File: examples/guests/csharp/validator/Validator.cs
using System.Runtime.CompilerServices;
using Polyplug.Guest;
using Polyplug.Abi;

namespace Validator;

public sealed class ValidatorPlugin : IPipelineValidatorGuestContract
{
    // Host handle for this runtime, captured at instance creation.
    private readonly IntPtr _host;

    public ValidatorPlugin(IntPtr host)
    {
        _host = host;
    }

    public StringView Validate(StringView input)
    {
        string s = StringViewHelper.StripPrefix(input, "DECODED:");
        string[] parts = s.Split('|');
        if (parts.Length == 3 && !string.IsNullOrEmpty(parts[0]) && !string.IsNullOrEmpty(parts[1]) && int.TryParse(parts[2], out _))
        {
            return PolyplugHost.AllocString(_host, $"VALID:{s}");
        }
        return PolyplugHost.AllocString(_host, "INVALID:expected name|value|count");
    }
}

public static class Registration
{
    [ModuleInitializer]
    public static void Register()
    {
        // The factory receives the HostApi pointer per created instance.
        ValidatorInterfaces.SetValidatorFactory(host => new ValidatorPlugin(host));
    }
}

File: examples/guests/python/validator/validator.py
from generated.guest.contracts import (
    VALIDATORPipelineValidatorPlugin,
    set_validator_factory,
    polyplug_init,
)


class ValidatorImpl(VALIDATORPipelineValidatorPlugin):
    """The factory receives the HostApi pointer at polyplug_init time."""

    def __init__(self, host_ptr: int) -> None:
        # Host handle for this runtime, captured at construction.
        self._host_ptr: int = host_ptr

    def validate(self, input: str) -> str:
        s = input.removeprefix("DECODED:")
        parts = s.split("|")
        if len(parts) >= 3 and parts[0] and parts[1]:
            try:
                int(parts[2])
                return f"VALID:{s}"
            except ValueError:
                pass
        return "INVALID:expected format is name|value|count"


# Register the factory; the generated polyplug_init constructs the
# implementation with its owning runtime's host pointer.
set_validator_factory(ValidatorImpl)
